Behavior for directories with `0o000` permissions does not match native filesystems

**Describe the bug**
The behavior of file operations on directories with `0o000` permissions does not match the behavior of native filesystems for at least rmdir and chmod operations. The owner of a file/dir with `0o000` perms should always be able to remove or chmod them again, but pyfakefs does not respect this.

I've attached both a pytest test suite and bash script that demonstrates these operations and confirmed the behavior is the same on Ubuntu 18.04 and MacOS 12.5.1.

**How To Reproduce**
Here's a set of Pytest cases that show the operations passing natively and failing when run in pyfakefs. You can also directly run the python test file without using pytest to get an annotated output of the test.
<details>
<summary>Python test case</summary>

```py
import pyfakefs
import pytest

from pathlib import Path
from stat import S_IMODE

def get_path_perms(p):
    return S_IMODE(p.stat().st_mode)

def do_mkdir():
    path = Path("/tmp/test-chmod")

    print("Creating test directory set setting permissions to 000")
    path.mkdir(mode=0o000)
    print(f"{path.resolve()} has permissions {oct(get_path_perms(path))}")
    assert get_path_perms(path) == 0o000
    print()

    print("Listing test dir - expect to get permissions denied")
    with pytest.raises(PermissionError):
        try:
            list(path.iterdir())
        except PermissionError as e:
            print(f"Got PermissionError: {e}")
            raise
    print()

    print("Chmod test dir to 600 - expect to pass and print new permissions")
    path.chmod(0o600)
    print(f"{path.resolve()} has permissions {oct(get_path_perms(path))}")
    # Should run without exception but fails under pyfakefs
    list(path.iterdir())
    assert get_path_perms(path) == 0o600
    print()

    print("Set permissions back to 000")
    path.chmod(0o000)
    print(f"{path.resolve()} has permissions {oct(get_path_perms(path))}")
    assert get_path_perms(path) == 0o000
    print()

    print("Remove test directory while permissions 000 - expect to pass and show no such directory")
    path.rmdir()
    print(f"{path.resolve()} exists? {path.exists()}")
    assert not path.exists()
    print()

def test_do_mkdir_native():
    do_mkdir()

def test_do_mkdir_fakefs(fs):
    do_mkdir()

if __name__ == "__main__":
    do_mkdir()

```
</details>

<details>
<summary>Full python test output</summary>
Notice that the `test_do_mkdir_native` test case passes, but the `test_do_mkdir_fakefs` test case fails.

```
jesse.cureton@jesse:~/debugging_logs/20220923_pyfakefs_bug_report$ ls -la
total 20
drwxrwxr-x 2 jesse.cureton jesse.cureton 4096 Sep 23 14:39 .
drwxrwxr-x 4 jesse.cureton jesse.cureton 4096 Sep 23 13:43 ..
-rwxrwxr-x 1 jesse.cureton jesse.cureton  608 Sep 23 13:55 bash_test.sh
-rw-rw-r-- 1 jesse.cureton jesse.cureton 1539 Sep 23 14:38 python_test.py
jesse.cureton@jesse:~/debugging_logs/20220923_pyfakefs_bug_report$ pytest-3
===================================================================================== test session starts ======================================================================================
platform linux -- Python 3.6.9, pytest-3.6.4, py-1.5.2, pluggy-0.6.0
rootdir: /home/jesse.cureton/debugging_logs/20220923_pyfakefs_bug_report, inifile:
plugins: pyfakefs-4.6.3
collected 2 items

python_test.py .F                                                                                                                                                                        [100%]

=========================================================================================== FAILURES ===========================================================================================
_____________________________________________________________________________________ test_do_mkdir_fakefs _____________________________________________________________________________________

fs = <pyfakefs.fake_filesystem.FakeFilesystem object at 0x7f2ed2c85400>

    def test_do_mkdir_fakefs(fs):
>       do_mkdir()

/home/jesse.cureton/debugging_logs/20220923_pyfakefs_bug_report/python_test.py:52:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/jesse.cureton/debugging_logs/20220923_pyfakefs_bug_report/python_test.py:29: in do_mkdir
    path.chmod(0o600)
/usr/lib/python3.6/pathlib.py:1266: in chmod
    ???
/home/jesse.cureton/.local/lib/python3.6/site-packages/pyfakefs/fake_pathlib.py:124: in chmod
    return pathobj.filesystem.chmod(str(pathobj), *args, **kwargs)
/home/jesse.cureton/.local/lib/python3.6/site-packages/pyfakefs/fake_filesystem.py:1415: in chmod
    file_object = self.resolve(path, follow_symlinks, allow_fd=True)
/home/jesse.cureton/.local/lib/python3.6/site-packages/pyfakefs/fake_filesystem.py:2292: in resolve
    file_path, check_read_perm), check_read_perm)
/home/jesse.cureton/.local/lib/python3.6/site-packages/pyfakefs/fake_filesystem.py:2239: in get_object_from_normpath
    self.raise_os_error(errno.EACCES, target.path)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pyfakefs.fake_filesystem.FakeFilesystem object at 0x7f2ed2c85400>, err_no = 13, filename = '/tmp/test-chmod', winerror = None

    def raise_os_error(self, err_no: int,
                       filename: Optional[AnyString] = None,
                       winerror: Optional[int] = None) -> NoReturn:
        """Raises OSError.
            The error message is constructed from the given error code and shall
            start with the error string issued in the real system.
            Note: this is not true under Windows if winerror is given - in this
            case a localized message specific to winerror will be shown in the
            real file system.

            Args:
                err_no: A numeric error code from the C variable errno.
                filename: The name of the affected file, if any.
                winerror: Windows only - the specific Windows error code.
            """
        message = os.strerror(err_no) + ' in the fake filesystem'
        if (winerror is not None and sys.platform == 'win32' and
                self.is_windows_fs):
            raise OSError(err_no, message, filename, winerror)
>       raise OSError(err_no, message, filename)
E       PermissionError: [Errno 13] Permission denied in the fake filesystem: '/tmp/test-chmod'

/home/jesse.cureton/.local/lib/python3.6/site-packages/pyfakefs/fake_filesystem.py:1122: PermissionError
------------------------------------------------------------------------------------- Captured stdout call -------------------------------------------------------------------------------------
Creating test directory set setting permissions to 000
/tmp/test-chmod has permissions 0o0

Listing test dir - expect to get permissions denied
Got PermissionError: [Errno 13] Permission denied in the fake filesystem: '/tmp/test-chmod'

Chmod test dir to 600 - expect to pass and print new permissions
============================================================================== 1 failed, 1 passed in 0.32 seconds ==============================================================================
```
</details>

As a bonus, here's a Bash script that runs the same test to confirm that this does behave as expected. I've confirmed this on both MacOS and Linux.
<details>
<summary>Bash test case</summary>

```bash
#!/bin/bash

set +euo pipefail

cd /tmp

echo "Creating test directory set setting permissions to 000"
mkdir test-chmod
chmod 000 test-chmod
ls -la . | grep test-chmod
echo

echo "Listing test dir - expect to get permissions denied"
ls -la test-chmod/
echo

echo "Chmod test dir to 600 - expect to pass and print new permissions"
chmod 600 test-chmod
ls -la . | grep test-chmod
echo

echo "Set permissions back to 000"
chmod 000 test-chmod
ls -la . | grep test-chmod
echo

echo "Remove test directory while permissions 000 - expect to pass and show no such directory"
rmdir test-chmod
ls -la test-chmod
echo

```
 </details>

**Your environment**
On MacOS:
```
jesse.cureton@Jesses-MacBook-Pro:/tmp$ python3 -c "import platform; print(platform.platform())"
macOS-12.5.1-x86_64-i386-64bit
jesse.cureton@Jesses-MacBook-Pro:/tmp$ python3 -c "import sys; print('Python', sys.version)"
Python 3.9.9 (main, Nov 21 2021, 03:23:42) 
[Clang 13.0.0 (clang-1300.0.29.3)]
jesse.cureton@Jesses-MacBook-Pro:/tmp$ python3 -c "from pyfakefs.fake_filesystem import __version__; print('pyfakefs', __version__)"
pyfakefs 4.7.0
```
On Ubuntu:
```
jesse.cureton@jesse:~/debugging_logs/20220923_pyfakefs_bug_report$ python3 -c "import platform; print(platform.platform())"
Linux-5.4.0-1083-aws-x86_64-with-Ubuntu-18.04-bionic
jesse.cureton@jesse:~/debugging_logs/20220923_pyfakefs_bug_report$ python3 -c "import sys; print('Python', sys.version)"
Python 3.6.9 (default, Jun 29 2022, 11:45:57)
[GCC 8.4.0]
jesse.cureton@jesse:~/debugging_logs/20220923_pyfakefs_bug_report$ python3 -c "from pyfakefs.fake_filesystem import __version__; print('pyfakefs', __version__)"
pyfakefs 4.6.3
```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -10,6 +10,7 @@ The released versions correspond to PyPi releases.
 ### Fixes
 * fixed handling of `O_TMPFILE` in `os.open` (caused handling of 
   `O_DIRECTORY` as `O_TMPFILE`) (see [#723](../../issues/723))
+* fixed handling of read permissions (see [#719](../../issues/719))
 
 ## [Version 4.7.0](https://pypi.python.org/pypi/pyfakefs/4.7.0) (2022-09-18)
 Changed handling of nested fixtures and bug fixes.

--- a/pyfakefs/fake_filesystem.py
+++ b/pyfakefs/fake_filesystem.py
@@ -1413,7 +1413,8 @@ class FakeFilesystem:
             follow_symlinks: If `False` and `path` points to a symlink,
                 the link itself is affected instead of the linked object.
         """
-        file_object = self.resolve(path, follow_symlinks, allow_fd=True)
+        file_object = self.resolve(path, follow_symlinks, allow_fd=True,
+                                   check_owner=True)
         if self.is_windows_fs:
             if mode & PERM_WRITE:
                 file_object.st_mode = file_object.st_mode | 0o222
@@ -2199,7 +2200,8 @@ class FakeFilesystem:
 
     def get_object_from_normpath(self,
                                  file_path: AnyPath,
-                                 check_read_perm: bool = True) -> AnyFile:
+                                 check_read_perm: bool = True,
+                                 check_owner: bool = False) -> AnyFile:
         """Search for the specified filesystem object within the fake
         filesystem.
 
@@ -2208,6 +2210,9 @@ class FakeFilesystem:
                 path that has already been normalized/resolved.
             check_read_perm: If True, raises OSError if a parent directory
                 does not have read permission
+            check_owner: If True, and check_read_perm is also True,
+                only checks read permission if the current user id is
+                different from the file object user id
 
         Returns:
             The FakeFile object corresponding to file_path.
@@ -2236,11 +2241,21 @@ class FakeFilesystem:
                     self.raise_os_error(errno.ENOENT, path)
                 target = target.get_entry(component)  # type: ignore
                 if (not is_root() and check_read_perm and target and
-                        not target.st_mode & PERM_READ):
+                        not self._can_read(target, check_owner)):
                     self.raise_os_error(errno.EACCES, target.path)
         except KeyError:
             self.raise_os_error(errno.ENOENT, path)
         return target
+
+    @staticmethod
+    def _can_read(target, owner_can_read):
+        if target.st_uid == USER_ID:
+            if owner_can_read or target.st_mode & 0o400:
+                return True
+        if target.st_gid == GROUP_ID:
+            if target.st_mode & 0o040:
+                return True
+        return target.st_mode & 0o004
 
     def get_object(self, file_path: AnyPath,
                    check_read_perm: bool = True) -> FakeFile:
@@ -2265,7 +2280,8 @@ class FakeFilesystem:
     def resolve(self, file_path: AnyStr,
                 follow_symlinks: bool = True,
                 allow_fd: bool = False,
-                check_read_perm: bool = True) -> FakeFile:
+                check_read_perm: bool = True,
+                check_owner: bool = False) -> FakeFile:
         """Search for the specified filesystem object, resolving all links.
 
         Args:
@@ -2275,6 +2291,9 @@ class FakeFilesystem:
             allow_fd: If `True`, `file_path` may be an open file descriptor
             check_read_perm: If True, raises OSError if a parent directory
                 does not have read permission
+            check_owner: If True, and check_read_perm is also True,
+                only checks read permission if the current user id is
+                different from the file object user id
 
         Returns:
           The FakeFile object corresponding to `file_path`.
@@ -2290,7 +2309,7 @@ class FakeFilesystem:
 
         if follow_symlinks:
             return self.get_object_from_normpath(self.resolve_path(
-                file_path, check_read_perm), check_read_perm)
+                file_path, check_read_perm), check_read_perm, check_owner)
         return self.lresolve(file_path)
 
     def lresolve(self, path: AnyPath) -> FakeFile:
@@ -3299,13 +3318,17 @@ class FakeFilesystem:
         """
         return self._is_of_type(path, S_IFLNK, follow_symlinks=False)
 
-    def confirmdir(self, target_directory: AnyStr) -> FakeDirectory:
+    def confirmdir(
+            self, target_directory: AnyStr, check_owner: bool = False
+    ) -> FakeDirectory:
         """Test that the target is actually a directory, raising OSError
         if not.
 
         Args:
             target_directory: Path to the target directory within the fake
                 filesystem.
+            check_owner: If True, only checks read permission if the current
+                user id is different from the file object user id
 
         Returns:
             The FakeDirectory object corresponding to target_directory.
@@ -3313,7 +3336,8 @@ class FakeFilesystem:
         Raises:
             OSError: if the target is not a directory.
         """
-        directory = cast(FakeDirectory, self.resolve(target_directory))
+        directory = cast(FakeDirectory, self.resolve(
+            target_directory, check_owner=check_owner))
         if not directory.st_mode & S_IFDIR:
             self.raise_os_error(errno.ENOTDIR, target_directory, 267)
         return directory
@@ -3379,14 +3403,14 @@ class FakeFilesystem:
             self.raise_os_error(error_nr, target_directory)
         ends_with_sep = self.ends_with_path_separator(target_directory)
         target_directory = self.absnormpath(target_directory)
-        if self.confirmdir(target_directory):
+        if self.confirmdir(target_directory, check_owner=True):
             if not self.is_windows_fs and self.islink(target_directory):
                 if allow_symlink:
                     return
                 if not ends_with_sep or not self.is_macos:
                     self.raise_os_error(errno.ENOTDIR, target_directory)
 
-            dir_object = self.resolve(target_directory)
+            dir_object = self.resolve(target_directory, check_owner=True)
             if dir_object.entries:
                 self.raise_os_error(errno.ENOTEMPTY, target_directory)
             self.remove_object(target_directory)
@@ -4797,8 +4821,7 @@ class FakeOsModule:
         path = self._path_with_dir_fd(path, self.chown, dir_fd)
         file_object = self.filesystem.resolve(
             path, follow_symlinks, allow_fd=True)
-        if not ((is_int_type(uid) or uid is None) and
-                (is_int_type(gid) or gid is None)):
+        if not isinstance(uid, int) or not isinstance(gid, int):
             raise TypeError("An integer is required")
         if uid != -1:
             file_object.st_uid = uid

--- a/pyfakefs/tests/fake_os_test.py
+++ b/pyfakefs/tests/fake_os_test.py
@@ -25,7 +25,9 @@ import unittest
 from pyfakefs.helpers import IN_DOCKER, IS_PYPY
 
 from pyfakefs import fake_filesystem
-from pyfakefs.fake_filesystem import FakeFileOpen, is_root
+from pyfakefs.fake_filesystem import (
+    FakeFileOpen, is_root, USER_ID, set_uid, GROUP_ID, set_gid
+)
 from pyfakefs.extra_packages import (
     use_scandir, use_scandir_package, use_builtin_scandir
 )
@@ -5273,7 +5275,66 @@ class FakeOsUnreadableDirTest(FakeOsModuleTestBase):
         else:
             self.assertEqual(['some_file'], self.os.listdir(self.dir_path))
 
+    def test_listdir_user_readable_dir(self):
+        self.os.chmod(self.dir_path, 0o600)
+        self.assertEqual(['some_file'], self.os.listdir(self.dir_path))
+        self.os.chmod(self.dir_path, 0o000)
+
+    def test_listdir_user_readable_dir_from_other_user(self):
+        self.skip_real_fs()  # won't change user in real fs
+        user_id = USER_ID
+        set_uid(user_id + 1)
+        dir_path = self.make_path('dir1')
+        self.create_dir(dir_path, perm=0o600)
+        self.assertTrue(self.os.path.exists(dir_path))
+        set_uid(user_id)
+        if not is_root():
+            with self.assertRaises(PermissionError):
+                self.os.listdir(dir_path)
+        else:
+            self.assertEqual(['some_file'], self.os.listdir(self.dir_path))
+
+    def test_listdir_group_readable_dir_from_other_user(self):
+        self.skip_real_fs()  # won't change user in real fs
+        user_id = USER_ID
+        set_uid(user_id + 1)
+        dir_path = self.make_path('dir1')
+        self.create_dir(dir_path, perm=0o660)
+        self.assertTrue(self.os.path.exists(dir_path))
+        set_uid(user_id)
+        self.assertEqual([], self.os.listdir(dir_path))
+
+    def test_listdir_group_readable_dir_from_other_group(self):
+        self.skip_real_fs()  # won't change user in real fs
+        group_id = GROUP_ID
+        set_gid(group_id + 1)
+        dir_path = self.make_path('dir1')
+        self.create_dir(dir_path, perm=0o060)
+        self.assertTrue(self.os.path.exists(dir_path))
+        set_gid(group_id)
+        if not is_root():
+            with self.assertRaises(PermissionError):
+                self.os.listdir(dir_path)
+        else:
+            self.assertEqual([], self.os.listdir(dir_path))
+
+    def test_listdir_other_readable_dir_from_other_group(self):
+        self.skip_real_fs()  # won't change user in real fs
+        group_id = GROUP_ID
+        set_gid(group_id + 1)
+        dir_path = self.make_path('dir1')
+        self.create_dir(dir_path, perm=0o004)
+        self.assertTrue(self.os.path.exists(dir_path))
+        set_gid(group_id)
+        self.assertEqual([], self.os.listdir(dir_path))
+
     def test_stat_unreadable_dir(self):
+        self.assertEqual(0, self.os.stat(self.dir_path).st_mode & 0o666)
+
+    def test_chmod_unreadable_dir(self):
+        self.os.chmod(self.dir_path, 0o666)
+        self.assertEqual(0o666, self.os.stat(self.dir_path).st_mode & 0o666)
+        self.os.chmod(self.dir_path, 0o000)
         self.assertEqual(0, self.os.stat(self.dir_path).st_mode & 0o666)
 
     def test_stat_file_in_unreadable_dir(self):
@@ -5282,6 +5343,29 @@ class FakeOsUnreadableDirTest(FakeOsModuleTestBase):
                 errno.EACCES, self.os.stat, self.file_path)
         else:
             self.assertEqual(0, self.os.stat(self.file_path).st_size)
+
+    def test_remove_unreadable_dir(self):
+        dir_path = self.make_path('dir1')
+        self.create_dir(dir_path, perm=0o000)
+        self.assertTrue(self.os.path.exists(dir_path))
+        self.os.rmdir(dir_path)
+        self.assertFalse(self.os.path.exists(dir_path))
+
+    def test_remove_unreadable_dir_from_other_user(self):
+        self.skip_real_fs()  # won't change user in real fs
+        user_id = USER_ID
+        set_uid(user_id + 1)
+        dir_path = self.make_path('dir1')
+        self.create_dir(dir_path, perm=0o000)
+        self.assertTrue(self.os.path.exists(dir_path))
+        set_uid(user_id)
+        if not is_root():
+            with self.assertRaises(PermissionError):
+                self.os.rmdir(dir_path)
+            self.assertTrue(self.os.path.exists(dir_path))
+        else:
+            self.os.rmdir(dir_path)
+            self.assertFalse(self.os.path.exists(dir_path))
 
 
 class RealOsUnreadableDirTest(FakeOsUnreadableDirTest):

--- a/pyfakefs/tests/test_utils.py
+++ b/pyfakefs/tests/test_utils.py
@@ -299,10 +299,12 @@ class RealFsTestMixin:
         args = [to_string(arg) for arg in args]
         return self.os.path.join(self.base_path, *args)
 
-    def create_dir(self, dir_path):
+    def create_dir(self, dir_path, perm=0o777):
         """Create the directory at `dir_path`, including subdirectories.
         `dir_path` shall be composed using `make_path()`.
         """
+        if not dir_path:
+            return
         existing_path = dir_path
         components = []
         while existing_path and not self.os.path.exists(existing_path):
@@ -317,6 +319,7 @@ class RealFsTestMixin:
             existing_path = self.os.path.join(existing_path, component)
             self.os.mkdir(existing_path)
             self.os.chmod(existing_path, 0o777)
+        self.os.chmod(dir_path, perm)
 
     def create_file(self, file_path, contents=None, encoding=None, perm=0o666):
         """Create the given file at `file_path` with optional contents,

